File: scripts/demo/capture-phase10a-screenshots.mjs
#!/usr/bin/env node
/**
 * Phase 10A — capture real dashboard screenshots via Playwright/Chromium.
 * Requires core profile running and live telemetry warm-up.
 */
import { chromium } from "playwright";
import { execSync } from "node:child_process";
import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(__dirname, "../..");
const DASHBOARD_URL = process.env.DASHBOARD_URL || "http://localhost:3000";
const API_BASE = process.env.API_BASE || "http://localhost:8000";
const WARMUP_MS = Number(process.env.SCREENSHOT_WARMUP_MS || 30000);
const TIMESTAMP = new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19);
const OUT_DIR = join(ROOT, "docs/evidence/phase10/demo/screenshots", TIMESTAMP);
const LATEST_DIR = join(ROOT, "docs/evidence/phase10/demo/screenshots/latest");

function git(cmd) {
  try {
    return execSync(`git -C "${ROOT}" ${cmd}`, { encoding: "utf8" }).trim();
  } catch {
    return "unknown";
  }
}

function dockerProfiles() {
  try {
    const ps = execSync("docker compose --profile core ps --format json", {
      encoding: "utf8",
      cwd: ROOT,
    });
    const running = ps
      .split("\n")
      .filter(Boolean)
      .map((line) => JSON.parse(line))
      .filter((c) => c.State === "running")
      .map((c) => c.Service);
    return running.length ? `core (${running.join(", ")})` : "core (not running)";
  } catch {
    return "core (unknown)";
  }
}

const SHOTS = [
  {
    file: "00_dashboard_overview.png",
    testId: "connection-status",
    label: "Dashboard overview / connection status",
    fullPage: false,
  },
  {
    file: "01_live_telemetry_streams.png",
    testId: "live-telemetry",
    label: "Live telemetry streams",
    waitFor: "#event-counter",
    minEvents: 5,
  },
  {
    file: "02_edge_inference_and_fusion.png",
    testId: "edge-inference",
    label: "Edge inference / model scores",
    waitFor: "#model-score-counter",
    minEvents: 3,
  },
  {
    file: "03_agent_traces_and_hitl.png",
    testId: "agent-traces",
    label: "Agent traces and HITL context",
    before: async (page) => {
      await page.locator('[data-testid="agent-traces"]').scrollIntoViewIfNeeded();
    },
  },
  {
    file: "04_digital_twin_state_mirror.png",
    testId: "digital-twin",
    label: "Digital twin state mirror",
  },
  {
    file: "05_evidence_center_or_observability.png",
    testId: "operational-status",
    label: "Operational status / evidence observability",
    extra: async (page) => {
      await page.locator('[data-testid="mission-control"]').scrollIntoViewIfNeeded();
    },
  },
  {
    file: "06_failure_or_degraded_mode_if_available.png",
    testId: "failure-injection",
    label: "Failure injection (simulated degraded mode)",
    before: async (page) => {
      const btn = page.locator('button[data-scenario="sensor_dropout"]');
      if (await btn.isVisible()) {
        await btn.click();
        await page.waitForTimeout(2000);
      }
    },
  },
  {
    file: "07_ros2_nav_slam_compose_status_if_available.png",
    testId: "nav-slam",
    label: "ROS2 Nav2/SLAM MiniLab (compose-validated; offline in core-only)",
  },
];

async function waitForCounter(page, selector, min) {
  const deadline = Date.now() + WARMUP_MS;
  while (Date.now() < deadline) {
    const text = await page.locator(selector).textContent();
    const n = Number.parseInt(text || "0", 10);
    if (n >= min) return n;
    await page.waitForTimeout(1000);
  }
  throw new Error(`Timeout waiting for ${selector} >= ${min}`);
}

async function copyLatest(srcDir) {
  mkdirSync(LATEST_DIR, { recursive: true });
  for (const shot of SHOTS) {
    execSync(`cp "${join(srcDir, shot.file)}" "${join(LATEST_DIR, shot.file)}"`);
  }
}

async function main() {
  mkdirSync(OUT_DIR, { recursive: true });

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    viewport: { width: 1440, height: 900 },
    deviceScaleFactor: 1,
  });
  const page = await context.newPage();

  const results = [];
  let failed = false;

  try {
    await page.goto(DASHBOARD_URL, { waitUntil: "networkidle", timeout: 60000 });
    await page.waitForSelector('[data-testid="connection-status"]', { timeout: 30000 });

    // Warm-up: wait for live sensor + model score traffic
    try {
      await waitForCounter(page, "#event-counter", 5);
      await waitForCounter(page, "#model-score-counter", 3);
    } catch (err) {
      console.warn("Warm-up warning:", err.message);
      results.push({ note: `warm-up partial: ${err.message}` });
    }

    for (const shot of SHOTS) {
      const dest = join(OUT_DIR, shot.file);
      try {
        if (shot.before) await shot.before(page);
        const section = page.locator(`[data-testid="${shot.testId}"]`);
        await section.scrollIntoViewIfNeeded();
        await page.waitForTimeout(500);
        if (shot.waitFor && shot.minEvents) {
          await waitForCounter(page, shot.waitFor, shot.minEvents);
        }
        if (shot.extra) await shot.extra(page);

        if (shot.fullPage === false) {
          await section.screenshot({ path: dest });
        } else {
          await page.screenshot({ path: dest, fullPage: true });
        }
        results.push({ file: shot.file, status: "ok", label: shot.label });
        console.log(`Captured ${shot.file}`);
      } catch (err) {
        failed = true;
        results.push({ file: shot.file, status: "fail", error: err.message });
        console.error(`Failed ${shot.file}:`, err.message);
      }
    }
  } finally {
    await browser.close();
  }

  const metadata = {
    timestamp: TIMESTAMP,
    git_sha: git("rev-parse HEAD"),
    branch: git("branch --show-current"),
    command: "cd scripts/demo && npm run capture",
    dashboard_url: DASHBOARD_URL,
    api_base: API_BASE,
    docker_profiles: dockerProfiles(),
    warmup_ms: WARMUP_MS,
    screenshots: results,
    pass: !failed,
    notes: [
      "ROS2/Nav2/SLAM panel shows offline state under core-only profile — compose-validated, not live-gated.",
      "Screenshots are from real local execution; no generated or stock imagery.",
    ],
  };

  writeFileSync(join(OUT_DIR, "capture-metadata.json"), JSON.stringify(metadata, null, 2));
  copyLatest(OUT_DIR);

  console.log(`\nScreenshots saved to ${OUT_DIR}`);
  console.log(`Latest symlink copies in ${LATEST_DIR}`);

  if (failed) process.exit(1);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/demo/capture_phase10a_screenshots.py
#!/usr/bin/env python3
"""Phase 10A — capture real dashboard screenshots via Playwright/Chromium."""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
DASHBOARD_URL = os.environ.get("DASHBOARD_URL", "http://localhost:3000")
API_BASE = os.environ.get("API_BASE", "http://localhost:8000")
WARMUP_MS = int(os.environ.get("SCREENSHOT_WARMUP_MS", "30000"))
TIMESTAMP = datetime.now(timezone.utc).strftime("%Y%m%d-%H%M%S")
OUT_DIR = ROOT / "docs/evidence/phase10/demo/screenshots" / TIMESTAMP
LATEST_DIR = ROOT / "docs/evidence/phase10/demo/screenshots/latest"

SHOTS = [
    ("00_dashboard_overview.png", "connection-status", "Dashboard overview / connection status"),
    ("01_live_telemetry_streams.png", "live-telemetry", "Live telemetry streams"),
    ("02_edge_inference_and_fusion.png", "edge-inference", "Edge inference / model scores"),
    ("03_agent_traces_and_hitl.png", "agent-traces", "Agent traces and HITL context"),
    ("04_digital_twin_state_mirror.png", "digital-twin", "Digital twin state mirror"),
    ("05_evidence_center_or_observability.png", "operational-status", "Operational status / evidence"),
    ("06_failure_or_degraded_mode_if_available.png", "failure-injection", "Failure injection panel"),
    ("07_ros2_nav_slam_compose_status_if_available.png", "nav-slam", "Nav2/SLAM MiniLab offline panel"),
]


def git(cmd: str) -> str:
    try:
        return subprocess.check_output(
            ["git", "-C", str(ROOT), *cmd.split()],
            text=True,
        ).strip()
    except subprocess.CalledProcessError:
        return "unknown"


def docker_profiles() -> str:
    try:
        out = subprocess.check_output(
            ["docker", "compose", "--profile", "core", "ps", "--format", "json"],
            cwd=ROOT,
            text=True,
        )
        running = []
        for line in out.splitlines():
            if not line.strip():
                continue
            row = json.loads(line)
            if row.get("State") == "running":
                running.append(row.get("Service", "?"))
        return f"core ({', '.join(running)})" if running else "core (not running)"
    except (subprocess.CalledProcessError, json.JSONDecodeError, FileNotFoundError):
        return "core (unknown)"


def wait_for_counter(page, selector: str, minimum: int, deadline_ms: int) -> int:
    import time

    deadline = time.time() + deadline_ms / 1000
    while time.time() < deadline:
        text = page.locator(selector).text_content() or "0"
        value = int(text.strip() or "0")
        if value >= minimum:
            return value
        page.wait_for_timeout(1000)
    raise TimeoutError(f"Timeout waiting for {selector} >= {minimum}")


def main() -> int:
    try:
        from playwright.sync_api import sync_playwright
    except ImportError:
        print(
            "Playwright not installed. Run:\n"
            "  .venv/bin/pip install playwright\n"
            "  .venv/bin/playwright install chromium",
            file=sys.stderr,
        )
        return 1

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    results: list[dict] = []
    failed = False

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page(viewport={"width": 1440, "height": 900})
        page.goto(DASHBOARD_URL, wait_until="networkidle", timeout=60000)
        page.wait_for_selector('[data-testid="connection-status"]', timeout=30000)

        try:
            wait_for_counter(page, "#event-counter", 5, WARMUP_MS)
            wait_for_counter(page, "#model-score-counter", 3, WARMUP_MS)
        except TimeoutError as err:
            print(f"Warm-up warning: {err}")
            results.append({"note": f"warm-up partial: {err}"})

        for filename, test_id, label in SHOTS:
            dest = OUT_DIR / filename
            try:
                if test_id == "failure-injection":
                    btn = page.locator('button[data-scenario="sensor_dropout"]')
                    if btn.is_visible():
                        btn.click()
                        page.wait_for_timeout(2000)
                if test_id == "operational-status":
                    page.locator('[data-testid="mission-control"]').scroll_into_view_if_needed()

                section = page.locator(f'[data-testid="{test_id}"]')
                section.scroll_into_view_if_needed()
                page.wait_for_timeout(500)
                section.screenshot(path=str(dest))
                results.append({"file": filename, "status": "ok", "label": label})
                print(f"Captured {filename}")
            except Exception as err:  # noqa: BLE001
                failed = True
                results.append({"file": filename, "status": "fail", "error": str(err)})
                print(f"Failed {filename}: {err}", file=sys.stderr)

        browser.close()

    LATEST_DIR.mkdir(parents=True, exist_ok=True)
    for filename, _, _ in SHOTS:
        src = OUT_DIR / filename
        if src.exists():
            shutil.copy2(src, LATEST_DIR / filename)

    metadata = {
        "timestamp": TIMESTAMP,
        "git_sha": git("rev-parse HEAD"),
        "branch": git("branch --show-current"),
        "command": ".venv/bin/python scripts/demo/capture_phase10a_screenshots.py",
        "dashboard_url": DASHBOARD_URL,
        "api_base": API_BASE,
        "docker_profiles": docker_profiles(),
        "warmup_ms": WARMUP_MS,
        "screenshots": results,
        "pass": not failed,
        "notes": [
            "ROS2/Nav2/SLAM panel shows offline state under core-only profile — compose-validated.",
            "Screenshots from real local execution; no stock or generated imagery.",
        ],
    }
    (OUT_DIR / "capture-metadata.json").write_text(json.dumps(metadata, indent=2) + "\n")
    print(f"\nScreenshots saved to {OUT_DIR}")
    return 1 if failed else 0


if __name__ == "__main__":
    raise SystemExit(main())
